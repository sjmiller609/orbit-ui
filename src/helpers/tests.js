import { BrowserRouter } from 'react-router-dom'
import { shape } from 'prop-types'
import { mount } from 'enzyme'

// Instantiate router context
const router = {
  history: new BrowserRouter().history,
  route: {
    location: {},
    match: {},
  },
}

const createContext = () => ({
  context: { router },
  childContextTypes: { router: shape({}) },
})

// wrap with router context - to mount RR links correctly
export function mountWrap(node) {
  return mount(node, createContext())
}
