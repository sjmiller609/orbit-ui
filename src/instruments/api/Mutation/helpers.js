const has = (str, find) => ~str.toLowerCase().indexOf(find.toLowerCase())

export const formErrors = ({ error, errors }) => {
  if (!error) return

  let err = JSON.stringify(error).toLowerCase()
  return errors.find(e => {
    if (e.key && has(err, e.key)) return e
    else if (e.allKeys && e.allKeys.every(k => has(err, k))) return e
    else if (e.someKeys && e.someKeys.some(k => has(err, k))) return e
  })
}
