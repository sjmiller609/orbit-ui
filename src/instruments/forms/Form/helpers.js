// faltten and stringify nested keys
export const unpack = (obj, ns, hash) => {
  if (!obj) return
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keypath = ns ? `${ns}.${key}` : key
    if (value && typeof value === 'object') {
      return unpack(value, keypath, acc)
    }
    acc[keypath] = value
    return acc
  }, hash || {})
}

const roll = (keys, obj, value) => {
  const key = keys[0]
  // case 1 - not obj
  if (keys.length === 1) {
    obj[key] = value
    return obj
  }
  // case 2 - doesn't exist
  if (!obj[key]) {
    obj[key] = isNaN(keys[1]) ? {} : []
  }
  // case 3 - exists
  obj[key] = roll(
    keys.slice(1),
    Array.isArray(obj[key]) ? [...obj[key]] : { ...obj[key] },
    value
  )

  return obj
}

// convert back into obj
export const pack = (obj, type = {}) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    const keys = key.split('.')

    return roll(keys, acc, value)
  }, type)
}

export const packChild = ({ name, obj }) => {
  const obj2 = {}
  let isArray
  Object.entries(obj).forEach(k => {
    if (name.length < k[0].length && k[0].indexOf(name) === 0) {
      const n = k[0].slice(name.length + 1)
      // must be true for all matching
      isArray = !isNaN(n.slice(0, n.indexOf('.')))
      obj2[n] = k[1]
    }
  })
  if (Object.keys(obj2).length) return pack(obj2, isArray ? [] : {})
  return null
}

export const removeChild = ({ name, obj = {} }) => {
  const obj2 = {}
  Object.entries(obj).forEach(k => {
    if (name.length >= k[0].length || k[0].indexOf(name) !== 0) {
      obj2[k[0]] = k[1]
    }
  })
  return obj2
}
