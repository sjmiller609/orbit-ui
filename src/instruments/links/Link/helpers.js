// check if url is external
export const externalUrl = url => {
  if (typeof url === 'object') return false
  const path = /^https?:\/\/|^\/\//i
  if (path.test(url) && !~url.indexOf(window.location.origin)) return true
  return
}
