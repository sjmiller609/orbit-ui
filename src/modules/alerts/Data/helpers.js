import { formErrors } from 'instruments'

const errors = [
  {
    key: 'permissions',
    name: 'properties.alert_emails',
    error: 'Contact your workspace admin to upgrade your permissions.',
  },
]

export const handleError = error => formErrors({ error, errors })

export const trimError = error => error.split(':')[1].trim()
