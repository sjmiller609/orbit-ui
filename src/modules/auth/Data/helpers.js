import React from 'react'

import { formErrors, Link } from 'instruments'

const errors = [
  {
    key: 'invalid password',
    name: 'password',
    error: 'Incorrect password',
  },
  {
    key: 'not found',
    name: 'email',
    error: "Hmm, we can't find that email",
  },
  {
    key: 'email already in use',
    name: 'email',
    error: 'That email is already taken.',
  },
  {
    key: 'no password credentials found',
    name: 'email',
    error: 'No password found. Did you mean to login with OAuth?',
  },
  {
    key: 'awaiting email confirmation',
    name: 'email',
    error: (
      <Link to="/resend">
        Your email is not yet verified. Please check your email for a
        verification link.<br /> [Click to resend]
      </Link>
    ),
  },
  {
    key: 'not associated with the specified invite',
    name: 'email',
    error: 'The email you entered does not match your invitation.',
  },
  {
    key: 'permissions',
    name: 'permissions',
    error: 'Contact your Workspace Admin to upgrade your permissions.',
  },
  {
    key: 'Email address in use',
    name: 'email',
    error: (
      <Link to="/login">
        A user with that email address already exists. Click here to login.
      </Link>
    ),
  },
]

export const handleError = error => formErrors({ error, errors })
