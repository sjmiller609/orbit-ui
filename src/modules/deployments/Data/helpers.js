import { formErrors } from 'instruments'

const errors = [
  {
    key: 'Workspace already has a deployment named',
    name: 'label',
    error: 'Workspace already has a deployment with that name.',
  },
  {
    allKeys: ['duplicate', 'label'],
    name: 'label',
    error: 'Workspace already has a deployment with that name.',
  },
]

export const handleError = error => formErrors({ error, errors })

export const trimError = error => error.split(':')[1].trim()
