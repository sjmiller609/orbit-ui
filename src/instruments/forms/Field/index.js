import PropTypes from 'prop-types'
import React from 'react'
import classnames from 'classnames'
import s from './styles.scss'
import { FormLabel } from 'instruments'
import { unCamelCase } from 'helpers/format'
import { jsonEqual, isEqual } from 'helpers/compare'

const Field = Component => {
  class Field extends React.Component {
    field = null
    timeout = null
    onChange = this.onChange.bind(this)
    validate = this.validate.bind(this)
    showError = this.showError.bind(this)
    id = this.props.id || this.props.name
    state = {
      showError: this.props.showError || false,
      touched: false,
    }
    componentWillMount() {
      const { defaultValue, value, convert } = this.props
      // add timeout so that multiple set states don't overwrite each other
      if (defaultValue && value === null) {
        const v = convert ? convert(defaultValue) : defaultValue
        setTimeout(() => this.onChange(null, v), 0)
      }
    }

    componentDidMount() {
      const { focus } = this.props
      if (focus) this.field.focus()
    }

    componentWillReceiveProps({ value, type, submitted, showError }) {
      const set = {}

      // for keyValue field
      if (showError !== this.props.showError) set.showError = showError

      // update type (for show password)
      if (type !== this.props.type) set.type = type

      // show error on submit
      if (submitted && !this.props.submitted) set.showError = true

      if (!jsonEqual(value, this.props.value)) {
        set.showError = false

        // show error after 3 seconds of stopped typing
        if (this.timeout) clearTimeout(this.timeout)
        this.timeout = setTimeout(() => {
          this.showError()
          clearTimeout(this.timeout)
        }, 3000)

        set.touched = true
      }

      this.setState(set)
    }

    // should probably switch this to pull the props that don't change, and then json.stringify
    // NOTE: if any new props are passed in that are meant to be reactive, won't work.
    shouldComponentUpdate(
      {
        value,
        error,
        type,
        submitted,
        name,
        showError: showError2,
        required,
        data,
      },
      { showError, touched }
    ) {
      if (!jsonEqual(value, this.props.value)) return true
      if (!isEqual(error, this.props.error)) return true
      if (!isEqual(type, this.props.type)) return true
      if (!isEqual(name, this.props.name)) return true
      if (!isEqual(submitted, this.props.submitted)) return true

      if (!isEqual(required, this.props.required)) return true
      // read from props also, but may not have error
      if (!isEqual(showError2, this.props.showError)) return true
      if (!isEqual(showError, this.state.showError)) return true
      if (!isEqual(touched, this.state.touched)) return true
      // some fields require original data to be passed through
      if (!isEqual(data, this.props.data)) return true
      return false
    }

    componentWillUnmount() {
      clearTimeout(this.timeout)
    }

    onChange(e, v) {
      const { name, onChange, convert } = this.props
      let value2 = e ? e.target.value : v
      if (convert) value2 = convert(value2, true)
      onChange(name, value2)
    }

    showError(showError = true) {
      if (showError === this.state.showError) return
      this.setState({ showError })
    }

    validate(value) {
      const { validate, name, required } = this.props
      if (required && (typeof value !== 'boolean' && !value)) {
        const n = name.split('.').pop()
        return unCamelCase(n) + ' is required'
      }
      if (value && validate) {
        return validate(value)
      }

      return null
    }

    render() {
      /*eslint-disable no-unused-vars */
      const {
        error,
        label,
        info,
        submitted,
        focus,
        required,
        className,
        convert,
        value,
        ...props
      } = this.props
      const { showError, touched } = this.state
      const err = showError && !!error
      const newProps = {
        ...props,
        className: classnames(
          s.field,
          err ? s.error : null,
          required ? s.required : null,
          className
        ),
        id: this.id,
        required,
        showError,
        setRef: ref => (this.field = ref),
        validate: this.validate,
        onChange: this.onChange,
        onBlur: touched ? () => this.showError(true) : null,
        setShowError: this.showError,
        error: err ? <div className={s.errorMsg}>{error}</div> : null,
        label: label ? (
          <FormLabel id={this.id} info={info}>
            {label}
          </FormLabel>
        ) : null,
      }
      if (value) newProps.value = value
      if (convert) {
        newProps.value = convert(newProps.value)
        newProps.convert = convert
      }

      return <Component {...newProps} />
    }
  }

  Field.propTypes = {
    type: PropTypes.string,
    name: PropTypes.string.isRequired,
    id: PropTypes.string,
    onChange: PropTypes.func.isRequired, // pass in key value update function
    validate: PropTypes.func, // check if isValid, return error message
    required: PropTypes.bool,
    focus: PropTypes.bool,
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    value: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.bool,
      PropTypes.object,
      PropTypes.array,
    ]),
    data: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.bool,
      PropTypes.object,
      PropTypes.array,
    ]),
    defaultValue: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.number,
      PropTypes.bool,
      PropTypes.object,
      PropTypes.array,
    ]),
    title: PropTypes.string,
    error: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
    className: PropTypes.string,
    updateErrors: PropTypes.func,
    submitted: PropTypes.bool,
    showError: PropTypes.bool,
    info: PropTypes.string,
    convert: PropTypes.func,
  }

  Field.defaultProps = {
    value: '',
    label: '',
  }

  return Field
}

export default Field
