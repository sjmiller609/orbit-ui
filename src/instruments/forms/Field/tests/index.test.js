import React from 'react'
import { mountWrap } from 'helpers/tests'
import Field from '../index'

const props = {
  type: 'text',
  name: 'textField',
  id: '1234',
  onChange: jest.fn(),
  validate: jest.fn(),
  required: true,
  focus: true,
  label: React.Fragment,
  value: '',
  data: '',
  defaultValue: '',
  title: 'Test',
  error: 'Not a valid address',
  className: 'testClassString',
  updateErrors: jest.fn(),
  submitted: false,
  showError: false,
  info: 'Type your email address',
  convert: jest.fn(),
}

describe('Field', () => {
  it('renders correctly', () => {
    const wrapper = mountWrap(<Field {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
