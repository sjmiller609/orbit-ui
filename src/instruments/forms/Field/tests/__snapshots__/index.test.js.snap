// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
<Field
  className="testClassString"
  error={
    <p>
      Error
    </p>
  }
  fieldId="testDeployType"
  id="1234"
  label={
    <label>
      Test
    </label>
  }
  name="testDeployType"
  onBlur={[Function]}
  onChange={[Function]}
  required={true}
  setRef={[Function]}
  title="Test Checkbox"
  type="radio"
  updateErrors={[Function]}
  validate={[Function]}
  value="Kubernetes"
>
  <Checkbox
    className="field required testClassString"
    error={null}
    fieldId="testDeployType"
    id="1234"
    label={
      <FormLabel
        id="1234"
        info={undefined}
      >
        <label>
          Test
        </label>
      </FormLabel>
    }
    name="testDeployType"
    onBlur={null}
    onChange={[Function]}
    required={true}
    setRef={[Function]}
    setShowError={[Function]}
    showError={false}
    title="Test Checkbox"
    type="radio"
    updateErrors={[Function]}
    validate={[Function]}
    value="Kubernetes"
  >
    <Row
      align="flex-start"
      auto={true}
      className="checkbox field required testClassString"
      id="testDeployType"
      justify="flex-start"
    >
      <Flex
        align="flex-start"
        auto={true}
        className="checkbox field required testClassString"
        flow="row"
        id="testDeployType"
        justify="flex-start"
      >
        <div
          className="flex row flex-start align-flex-start auto checkbox field required testClassString"
          id="testDeployType"
        >
          <input
            checked={true}
            className="input"
            name="testDeployType"
            onBlur={null}
            onChange={[Function]}
            required={true}
            title="Test Checkbox"
            type="checkbox"
            value="Kubernetes"
          />
          <div
            className="box on"
            onClick={[Function]}
          >
            <div
              className="check"
            />
          </div>
          <FormLabel
            id="1234"
          >
            <label
              className="label"
              htmlFor="1234"
            >
              <label>
                Test
              </label>
              <Info />
            </label>
          </FormLabel>
        </div>
      </Flex>
    </Row>
  </Checkbox>
</Field>
`;
